src/julia_gc.c

```diff
@@ -450,6 +452,10 @@ void InitMarkFuncBags(UInt type, TNumMarkFuncBags mark_func)
 
 static inline int JMark(void * obj)
 {
+    void *ty = jl_typeof(obj);
+    if (ty != datatype_mptr && ty != datatype_bag
+        && ty != datatype_largebag && ty != jl_weakref_type)
+        return 0;
```

Comment:
Hm, so that's quite a bit different from the changes we discussed last. So, this is a hard coded list of types -- so we are making the hard assumption here that GAP objects will *never* directly reference other Julia objects. But that is not only fragile, it is also quite simply wrong, because JuliaInterface of course does so.

Hence, I don't understand how this can work at all.

I also don't quite understand why it is now sufficient to perform this check in `JMark` alone -- is it safe to increment `YoungRef` in `MarkBag` when dealing with invalid bags?

Reply:
Hmm, I hadn't realized that JuliaInterface goes through `JMark()` also. This is something that needs to be fixed indeed. I understand the fragility concern, but this is a hot path, where we need to look up these types fast.

The `MarkBag()` case is safe. We have at this point already validated that `p` is a valid address. If on a freelist or if unmarked, then the lowest bit will be zero and we will not increment `YoungRef`.

I am still going to rewrite this for better readability (and, of course, to remove the `JMark()` problem.

Reply:
Ah, I somehow overlooked that `MarkBag` already validates the pointer (which it indeed does).